crates/iota-sdk/src/lib.rs

```diff
@@ -644,3 +636,90 @@ impl DataReader for ReadApi {
         Ok(self.get_reference_gas_price().await?)
     }
 }
+
+pub trait GetAllPages<O, C, F, E>: Sized + Fn(Option<C>) -> F
+where
+    O: Send,
+    C: Send,
+    F: futures::Future<Output = Result<Page<O, C>, E>> + Send,
+{
+    fn get_all_pages(self) -> impl futures::Future<Output = Result<Vec<O>, E>> {
+        self.get_all_pages_stream().try_collect()
+    }
+
+    fn get_all_pages_stream(self) -> GetAllPagesStream<O, C, F, E, Self> {
+        GetAllPagesStream::new(self)
+    }
+}
+
+impl<T, O, C, F, E> GetAllPages<O, C, F, E> for T
+where
+    T: Fn(Option<C>) -> F,
+    O: Send,
+    C: Send,
+    F: futures::Future<Output = Result<Page<O, C>, E>> + Send,
+{
+}
+
+pub struct GetAllPagesStream<O, C, F, E, Fun> {
+    cursor: Option<C>,
+    fun: Fun,
+    fut: Pin<Box<F>>,
+    next: VecDeque<O>,
+    has_next_page: bool,
+    _data: PhantomData<fn(E) -> E>,
+}
+
+impl<O, C, F, E, Fun> GetAllPagesStream<O, C, F, E, Fun>
+where
+    Fun: Fn(Option<C>) -> F,
+{
+    pub fn new(fun: Fun) -> Self {
+        let fut = fun(None);
+        Self {
+            cursor: None,
+            fun,
+            fut: Box::pin(fut),
+            next: Default::default(),
+            has_next_page: true,
+            _data: PhantomData,
+        }
+    }
+}
+
+impl<O, C, F, E, Fun> futures::Stream for GetAllPagesStream<O, C, F, E, Fun>
+where
+    O: Send,
+    C: Send,
+    F: futures::Future<Output = Result<Page<O, C>, E>> + Send,
+    Fun: Fn(Option<C>) -> F,
+{
+    type Item = Result<O, E>;
+
+    fn poll_next(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        let this = unsafe { self.get_unchecked_mut() };
+        if this.next.is_empty() && this.has_next_page {
+            match this.fut.as_mut().poll(cx) {
+                Poll::Ready(res) => match res {
+                    Ok(mut page) => {
+                        this.cursor = page.next_cursor.take();
+                        this.next.extend(page.data);
+                        this.has_next_page = page.has_next_page;
+                        if this.has_next_page {
+                            this.fut.set((this.fun)(this.cursor.take()));
```

Comment:
Can't you just do this?
```suggestion
                        this.next.extend(page.data);
                        this.has_next_page = page.has_next_page;
                        if this.has_next_page {
                            this.fut.set((this.fun)(page.next_cursor.take()));
```

Reply:
Well, you're right that the cursor doesn't need to be stored. But the rest is still needed as there's no way to tell just from the vector whether there's more data